ethcontract/src/log.rs

```diff
@@ -282,19 +297,8 @@ impl<T: Transport> PastLogsStream<T> {
             .unwrap_or(DEFAULT_BLOCK_PAGE_SIZE);
         let filter = builder.into_filter();
 
-        let start_block = match from_block {
-            BlockNumber::Earliest => Some(0),
-            BlockNumber::Number(value) => Some(value.as_u64()),
-            BlockNumber::Latest | BlockNumber::Pending => None,
-        };
-        let end_block = match to_block {
-            BlockNumber::Earliest => None,
-            BlockNumber::Number(value) => Some(value.as_u64()),
-            BlockNumber::Latest | BlockNumber::Pending => {
-                let latest_block = web3.eth().block_number().await?;
-                Some(latest_block.as_u64())
-            }
-        };
+        let start_block = block_number(&web3, from_block).await?;
+        let end_block = block_number(&web3, to_block).await?;
```

Comment:
Wasn't sure what the best way to handle the new BlockNumber::Finalized and BlockNumber::Safe was. So I moved all non number block numbers to be fetched by the node.

Reply:
This seems slightly incorrect to me.

For logs, `to_block` can be past the `latest`, so it shouldn't error in that case. I would implement this as:

```
        let start_block = match from_block {
            BlockNumber::Earliest => Some(0),
            BlockNumber::Number(value) => Some(value.as_u64()),
            BlockNumber::Latest | BlockNumber::Pending => None,
            BlockNumber::Safe | BlockNumber::Finalized => block_number(&web3, from_block).await?,
        };
        let end_block = match to_block {
            BlockNumber::Earliest => None,
            BlockNumber::Number(value) => Some(value.as_u64()),
            BlockNumber::Latest | BlockNumber::Pending => {
                let latest_block = web3.eth().block_number().await?;
                Some(latest_block.as_u64())
            }
            BlockNumber::Safe | BlockNumber::Finalized => block_number(&web3, from_block).await?,
        };
```

Reply:
Changed to your suggestion.